lib/elixir/lib/kernel/typespec.ex

```diff
@@ -587,10 +591,22 @@ defmodule Kernel.Typespec do
     end
   end
 
+  defp typespec({:record, _meta, _args}, _vars, caller, _state) do
+    compile_error(caller, "bad record type")
+  end
+
   # Handle ranges
-  defp typespec({:.., meta, args}, vars, caller, state) do
-    {args, state} = Enum.map_reduce(args, state, &typespec(&1, vars, caller, &2))
-    {{:type, line(meta), :range, args}, state}
+  defp typespec({:.., meta, [left, right]}, vars, caller, state) do
+    {left, state} = typespec(left, vars, caller, state)
+    {right, state} = typespec(right, vars, caller, state)
+
+    case {:erl_eval.partial_eval(left), :erl_eval.partial_eval(right)} do
```

Comment:
I would prefer if we do not depend on erl_eval. Is there a reason why we need it?

Reply:
I was currently writing a comment related to this.

In short words, it's required for negative values. But the logic is quite simple, so we can roll our own implementation to "expand" the values.

Reply:
Let's roll our own then, especially because this seems to be private API (it is not publicly documented). :)